File: libgomp/testsuite/libgomp.oacc-c-c++-common/collapse-3.c
/* { dg-do run } */

#include <string.h>
#include <stdlib.h>
#include <stdio.h>

int
main (void)
{
  int i2, l = 0, r = 0;
  int a[3][3][3];

  memset (a, '\0', sizeof (a));
  #pragma acc parallel
  #pragma acc loop collapse(4 - 1)
    for (int i = 0; i < 2; i++)
      for (int j = 0; j < 2; j++)
	for (int k = 0; k < 2; k++)
	  a[i][j][k] = i + j * 4 + k * 16;
#pragma acc parallel
    {
      #pragma acc loop collapse(2) reduction(|:l)
	for (i2 = 0; i2 < 2; i2++)
	  for (int j = 0; j < 2; j++)
	    for (int k = 0; k < 2; k++)
	      if (a[i2][j][k] != i2 + j * 4 + k * 16)
		l += 1;
    }

    for (i2 = 0; i2 < 2; i2++)
      for (int j = 0; j < 2; j++)
	for (int k = 0; k < 2; k++)
	  if (a[i2][j][k] != i2 + j * 4 + k * 16)
	    r += 1;

  if (l != r)
    abort ();
  return 0;
}

File: libgomp/testsuite/libgomp.oacc-fortran/collapse-3.f90
! { dg-do run }

program collapse3
  integer :: a(3,3,3), k, kk, kkk, l, ll, lll
  !$acc parallel
  !$acc loop collapse(3)
    do 115 k=1,3
dokk: do kk=1,3
        do kkk=1,3
          a(k,kk,kkk) = 1
        enddo
      enddo dokk
115   continue
  !$acc end parallel
  if (any(a(1:3,1:3,1:3).ne.1)) STOP 1

  !$acc parallel
  !$acc loop collapse(3)
dol: do 120 l=1,3
doll: do ll=1,3
        do lll=1,3
          a(l,ll,lll) = 2
        enddo
      enddo doll
120 end do dol
  !$acc end parallel
  if (any(a(1:3,1:3,1:3).ne.2)) STOP 2
end program collapse3
